vhdl_lang/src/analysis/types.rs

```diff
@@ -0,0 +1,647 @@
+                            scope,
+                            &constraint.pos,
+                            elem_type.base(),
+                            &mut constraint.item,
+                            diagnostics,
+                        )?;
+                    }
+                } else {
+                    diagnostics.error(
+                        pos,
+                        format!(
+                            "Array constraint cannot be used for {}",
+                            base_type.describe()
+                        ),
+                    );
+                }
+            }
+            SubtypeConstraint::Range(ref mut range) => {
+                if base_type.is_scalar() {
+                    self.range_with_ttyp(scope, base_type.into(), range, diagnostics)?;
+                } else {
+                    diagnostics.error(
+                        pos,
+                        format!(
+                            "Scalar constraint cannot be used for {}",
+                            base_type.describe()
+                        ),
+                    );
+                }
+            }
+            SubtypeConstraint::Record(ref mut constraints) => {
+                if let Type::Record(region) = base_type.kind() {
+                    for constraint in constraints.iter_mut() {
+                        let ElementConstraint { ident, constraint } = constraint;
+                        let des = Designator::Identifier(ident.item.clone());
+                        if let Some(elem) = region.lookup(&des) {
+                            self.analyze_subtype_constraint(
+                                scope,
+                                &constraint.pos,
+                                elem.type_mark().base(),
+                                &mut constraint.item,
+                                diagnostics,
+                            )?;
+                        } else {
+                            diagnostics.push(Diagnostic::no_declaration_within(
+                                &base_type, &ident.pos, &des,
+                            ))
+                        }
+                    }
+                } else {
+                    diagnostics.error(
+                        pos,
+                        format!(
+                            "Record constraint cannot be used for {}",
+                            base_type.describe()
+                        ),
+                    );
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn analyze_subtype_indication(
+        &self,
+        scope: &Scope<'a>,
+        subtype_indication: &mut SubtypeIndication,
+        diagnostics: &mut dyn DiagnosticHandler,
+    ) -> FatalResult {
+        if let Err(err) = self.resolve_subtype_indication(scope, subtype_indication, diagnostics) {
+            err.add_to(diagnostics)?;
+        }
+        Ok(())
+    }
+}
+
+pub(crate) fn find_full_type_definition<'a>(
```

Comment:
Suggestion: This function is tightly coupled with the ` analyze_declarative_part` function and should remain private within the `declarative.rs`